### PR TITLE
Introducing a custom HashMap to try to improve group by performance

### PR DESCRIPTION
**Pull request by zhang2014**
I hereby agree to the terms of the CLA available at: https://datafuse.rs/policies/cla/

## Summary

Experimental Pull Request.

## Changelog

- Improvement

### CODE REVIEW

**Comment by BohuTANG** [comment]
Cool.
This mod should in common/hashtables not query/common? :thinking:

**Comment by zhang2014** [comment]
> This mod should in common/hashtables not query/common? 🤔

HashTable will only be used in query. Maybe it should be in query? 
By the way, this is just an attempt to verify

**Comment by BohuTANG** [comment]
> > This mod should in common/hashtables not query/common? thinking
> 
> HashTable will only be used in query. Maybe it should be in query?

Ok, make sense.

> By the way, this is just an attempt to verify

**Comment by drmingdrmer** [comment]
Cool~

**Comment by drmingdrmer** [comment]
Out of curiosity, what is the advantage in this implementation that makes it faster than the std hash?

**Comment by jyizheng** [comment]
What is the collision resolution policy for this hashtable implementation?

**Comment by zhang2014** [comment]
> Out of curiosity, what is the advantage in this implementation that makes it faster than the std hash?

Current test, it is just a little better than std. I'm not sure whether we should switch to custom HashMap for this.

Optimize(from ClickHouse HashTable):
- Before 12582912 (2 ^ 23), the capacity is expanded by 4 times each time, otherwise it is expanded by 2 times
- Merge, get and insert operations (completed in one time, like std entry)
- Use custom hash algorithm(from murmurhash)
- Used std::alloc::realloc to resize

**Comment by PsiACE** [comment]
> Out of curiosity, what is the advantage in this implementation that makes it faster than the std hash?

std hash (siphash) is a not-so-fast hashing algorithm, the most common fast hash algorithm currently used by Rust is ahash for the general case. If the object is small, then fxhash and zwohash will be faster than ahash.

**Comment by zhang2014** [comment]
> What is the collision resolution policy for this hashtable implementation?

It's a open addressing hash table

**Comment by PsiACE** [comment]
>     * Before 12582912 (2 ^ 23), the capacity is expanded by 4 times each time, otherwise it is expanded by 2 times

It looks like this could alleviate the tail delay. But I think the impact is still visible.

**Comment by BohuTANG** [comment]
> > Out of curiosity, what is the advantage in this implementation that makes it faster than the std hash?
> 
> Current test, it is just a little better than std. I'm not sure whether we should switch to custom HashMap for this.

:+1: , let's do it later.

> Optimize(from ClickHouse HashTable):
> 
> * Before 12582912 (2 ^ 23), the capacity is expanded by 4 times each time, otherwise it is expanded by 2 times
> * Merge, get and insert operations (completed in one time, like std entry)
> * Use custom hash algorithm(from murmurhash)
> * Used std::alloc::realloc to resize

**Comment by zhang2014** [comment]
> 👍 , let's do it later.

Ok. Let's temporarily close this pr

**Comment by zhang2014** [comment]
Reopen, performance test results from @sundy-li (without native compile flag):
```
datafuse : ) 60 rows in set. Elapsed: 1.583 sec. Processed 1.00 billion rows, 8.00 GB (631.70 million rows/s., 5.05 GB/s.)
clickhouse : ) 60 rows in set. Elapsed: 1.776 sec. Processed 1.00 billion rows, 8.00 GB (563.11 million rows/s., 4.50 GB/s.)
```

File: common/infallible/src/rwlock.rs [hunk 1]
```diff
@@ -11,6 +11,8 @@ use parking_lot::RwLockWriteGuard;
 #[derive(Debug, Default)]
 pub struct RwLock<T>(ParkingRwLock<T>);
 
+unsafe impl<T> Send for RwLock<T> {}
```
Comment (sundy-li) [inline common/infallible/src/rwlock.rs:14]: Should add `where T:Send` ?

File: query/src/common/hash_table_iter.rs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+                    return self.zero_entity;
+                }
+            }
+
+            self.idx += 1;
+            while self.idx < self.capacity && self.entities.offset(self.idx).is_zero() {
+                self.idx += 1;
+            }
+
+            match self.idx == self.capacity {
```
Comment (leiysky) [inline query/src/common/hash_table_iter.rs:45]: What about use `if-else` here?

**Comment by leiysky** [comment]
Seems it's a linear probing hash scheme, which can utilize with vectorization easily(by supporting `batch_get` and `batch_insert`). 

I've just found a [paper](https://www.researchgate.net/publication/326669722_SIMD_Vectorized_Hashing_for_Grouped_Aggregation_22nd_European_Conference_ADBIS_2018_Budapest_Hungary_September_2-5_2018_Proceedings) about this, FYI.

File: query/src/common/hash_table.rs [hunk 3]
```diff
@@ -0,0 +1,216 @@
+// Copyright 2020-2021 The Datafuse Authors.
```
Comment (BohuTANG) [inline query/src/common/hash_table.rs:1]: Aha, we have the new licenese header now :)

**Comment by zhang2014** [comment]
> Seems it's a linear probing hash scheme, which can utilize with vectorization easily(by supporting batch_get and batch_insert).
> 
> I've just found a paper about this, FYI.

This is really a great idea. Let's create a issue to record it.
By the way, in actual tests, using rust auto vectorization may be better than write SIMD instructions(I guess this may be because it can always improve the utilization of registers well enough)

**Review by sundy-li — APPROVED** [review]
LGTM

**Pull request merged**